### ISSUE TITLE
Signal 4.19.3 (non-Google version) crashes after update or fresh install/db restoration

### ISSUE THREAD

**Issue by anon1701** [signalapp/Signal-Android#7901]
- [X] I have searched open and closed issues for duplicates
- [X] I am submitting a bug report for existing functionality that does not work as intended
- [X] I have read https://github.com/signalapp/Signal-Android/wiki/Submitting-useful-bug-reports
- [X] This isn't a feature request or a discussion topic

----------------------------------------

### Bug description
I am using the non-Google Services version of Signal, apk downloaded from Signal's web pages.
I was using version 4.18.x, made a backup of my Signal data and installed (updated) the newest version (4.19.3) with the downloaded apk file.
After the update Signal crashed repeatedly at startup. I removed Signal completely and installed the version 4.19.3 apk again. I tried to restore the freshly made backup when Signal prompted at first start.

### Steps to reproduce
Using the non-Google version of Signal:
- Update version 4.18.x to 4.19.3.
- Try to start Signal.
- Signal crashes at startup.

- Remove Signal completely
- Restore backup
- Signal crashes when restoring the backup
- Signal crashes also at startup.

**Actual result:** Signal crashes after the update, even with fresh install with restored data.
**Expected result:** Update Signal to 4.19.3 and continue using the updated application normally.

### Device info

**Device:** Samsung Galaxy S6 Edge
**Android version:** LineageOS 7.1.2
**Signal version:** 4.19.3

### Debug log

This is the error that I continuously see in logcat via adb when trying so start Signal after the update / db restoration:
`net.sqlcipher.database.SQLiteException: Can't upgrade read-only database from version 6 to 7: /data/user/0/org.thoughtcrime.securesms/databases/signal.db`

**Comment by moxie0** [comment]
We can pretty much never do anything without a debug log. Please include a full log and reopen, thanks!

**Comment by anon1701** [comment]
Hello, here comes a bit more of the debug log. Hopefully this will contain more information, retrieved with "logcat | grep securesms".
Sorry for the the bad layout of log.

First, the installation of the .apk and first start of Signal 4.19.3, trying to restore backup:

06-15 19:28:42.905 23130 23130 I AppSecurityPermissions: Ignoring unknown permission:org.thoughtcrime.securesms.ACCESS_SECRETS
06-15 19:28:42.924 23130 23130 I AppSecurityPermissions: Ignoring unknown permission:org.thoughtcrime.securesms.permission.C2D_MESSAGE
06-15 19:28:48.071  4893  4943 W PackageManager: Package org.thoughtcrime.securesms desires unavailable shared library com.sec.android.app.multiwindow; ignoring!
06-15 19:28:48.077  4893  4943 I PackageManager.DexOptimizer: Running dexopt (dex2oat) on: /data/app/vmdl1352777626.tmp/base.apk pkg=org.thoughtcrime.securesms isa=arm vmSafeMode=false debuggable=false target-filter=interpret-only oatDir = /data/app/vmdl1352777626.tmp/oat sharedLibraries=null
06-15 19:28:50.277  4893  4943 W PackageManager: Package org.thoughtcrime.securesms desires unavailable shared library com.sec.android.app.multiwindow; ignoring!
06-15 19:28:50.342  4893  4943 W PackageManager: Not granting permission android.permission.BROADCAST_WAP_PUSH to package org.thoughtcrime.securesms (protectionLevel=2 flags=0x38583e44)
06-15 19:28:50.343  4893  4943 W PackageManager: Unknown permission android.permission.READ_CALL_STATE in package org.thoughtcrime.securesms
06-15 19:28:50.343  4893  4943 W PackageManager: Unknown permission com.google.android.c2dm.permission.RECEIVE in package org.thoughtcrime.securesms
06-15 19:28:50.343  4893  4943 W PackageManager: Unknown permission android.permission.INSTALL_SHORTCUT in package org.thoughtcrime.securesms
06-15 19:28:50.343  4893  4943 W PackageManager: Unknown permission android.permission.RAISED_THREAD_PRIORITY in package org.thoughtcrime.securesms
06-15 19:28:50.343  4893  4943 W PackageManager: Unknown permission com.sec.android.provider.badge.permission.READ in package org.thoughtcrime.securesms
06-15 19:28:50.343  4893  4943 W PackageManager: Unknown permission com.sec.android.provider.badge.permission.WRITE in package org.thoughtcrime.securesms
06-15 19:28:50.343  4893  4943 W PackageManager: Unknown permission com.htc.launcher.permission.READ_SETTINGS in package org.thoughtcrime.securesms
06-15 19:28:50.343  4893  4943 W PackageManager: Unknown permission com.htc.launcher.permission.UPDATE_SHORTCUT in package org.thoughtcrime.securesms
06-15 19:28:50.343  4893  4943 W PackageManager: Unknown permission com.sonyericsson.home.permission.BROADCAST_BADGE in package org.thoughtcrime.securesms
06-15 19:28:50.343  4893  4943 W PackageManager: Unknown permission com.sonymobile.home.permission.PROVIDER_INSERT_BADGE in package org.thoughtcrime.securesms
06-15 19:28:50.343  4893  4943 W PackageManager: Unknown permission com.anddoes.launcher.permission.UPDATE_COUNT in package org.thoughtcrime.securesms
06-15 19:28:50.343  4893  4943 W PackageManager: Unknown permission com.majeur.launcher.permission.UPDATE_BADGE in package org.thoughtcrime.securesms
06-15 19:28:50.343  4893  4943 W PackageManager: Unknown permission com.huawei.android.launcher.permission.CHANGE_BADGE in package org.thoughtcrime.securesms
06-15 19:28:50.343  4893  4943 W PackageManager: Unknown permission com.huawei.android.launcher.permission.READ_SETTINGS in package org.thoughtcrime.securesms
06-15 19:28:50.343  4893  4943 W PackageManager: Unknown permission com.huawei.android.launcher.permission.WRITE_SETTINGS in package org.thoughtcrime.securesms
06-15 19:28:50.343  4893  4943 W PackageManager: Unknown permission android.permission.READ_APP_BADGE in package org.thoughtcrime.securesms
06-15 19:28:50.343  4893  4943 W PackageManager: Unknown permission com.oppo.launcher.permission.READ_SETTINGS in package org.thoughtcrime.securesms
06-15 19:28:50.343  4893  4943 W PackageManager: Unknown permission com.oppo.launcher.permission.WRITE_SETTINGS in package org.thoughtcrime.securesms
06-15 19:28:50.343  4893  4943 W PackageManager: Unknown permission me.everything.badger.permission.BADGE_COUNT_READ in package org.thoughtcrime.securesms
06-15 19:28:50.343  4893  4943 W PackageManager: Unknown permission me.everything.badger.permission.BADGE_COUNT_WRITE in package org.thoughtcrime.securesms
06-15 19:28:50.387  4640  4640 D installd: Detected label change from u:object_r:system_data_file:s0 to u:object_r:app_data_file:s0:c512,c768 at /data/data/org.thoughtcrime.securesms; running recursive restorecon
06-15 19:28:50.388  4640  4640 D installd: Detected label change from u:object_r:system_data_file:s0 to u:object_r:app_data_file:s0:c512,c768 at /data/user_de/0/org.thoughtcrime.securesms; running recursive restorecon
06-15 19:28:50.440  5048  5048 D BluetoothMapAppObserver: The installed package is: org.thoughtcrime.securesms
06-15 19:28:57.628 23346 23367 D InstalledAppProviderSer: insertAppIntoDb org.thoughtcrime.securesms
06-15 19:28:57.732 23346 23367 D InstalledAppProviderSer: insertAppIntoDb org.thoughtcrime.securesms
06-15 19:29:12.789  4893  5083 I ActivityManager: START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10000000 pkg=org.thoughtcrime.securesms cmp=org.thoughtcrime.securesms/.RoutingActivity} from uid 10024 on display 0
06-15 19:29:12.818  4893 11097 I ActivityManager: Start proc 23385:org.thoughtcrime.securesms/u0a113 for activity org.thoughtcrime.securesms/.RoutingActivity
06-15 19:29:13.120  4893  5083 D VoldConnector: SND -> {27 volume mkdirs /storage/emulated/0/Android/data/org.thoughtcrime.securesms/cache/}
06-15 19:29:13.457 23385 23385 W PersistentAlarmManagerListener: class org.thoughtcrime.securesms.service.RotateSignedPreKeyListener scheduling for: 1529252953456
06-15 19:29:13.470 23385 23385 W PersistentAlarmManagerListener: class org.thoughtcrime.securesms.service.DirectoryRefreshListener scheduling for: 1529123353470
06-15 19:29:13.474 23385 23385 W PersistentAlarmManagerListener: class org.thoughtcrime.securesms.service.UpdateApkRefreshListener scheduling for: 1529101753473
06-15 19:29:13.530  4893  5834 I ActivityManager: START u0 {cmp=org.thoughtcrime.securesms/.PassphraseCreateActivity (has extras)} from uid 10113 on display 0
06-15 19:29:13.851  4893  4937 I ActivityManager: Displayed org.thoughtcrime.securesms/.PassphraseCreateActivity: +242ms (total +1s46ms)
06-15 19:29:14.022  4893 11097 I ActivityManager: START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10000000 pkg=org.thoughtcrime.securesms cmp=org.thoughtcrime.securesms/.RoutingActivity} from uid 10113 on display 0
06-15 19:29:14.080  4893  5357 I ActivityManager: START u0 {cmp=org.thoughtcrime.securesms/.RegistrationActivity (has extras)} from uid 10113 on display 0
06-15 19:29:15.236  4893  4937 I ActivityManager: Displayed org.thoughtcrime.securesms/.RegistrationActivity: +1s132ms (total +1s188ms)
06-15 19:29:39.529 23385 23414 E AndroidRuntime: Process: org.thoughtcrime.securesms, PID: 23385
06-15 19:29:39.529 23385 23414 E AndroidRuntime: 	at org.thoughtcrime.securesms.database.helpers.SQLCipherOpenHelper.onUpgrade(SQLCipherOpenHelper.java:172)
06-15 19:29:39.529 23385 23414 E AndroidRuntime: 	at org.thoughtcrime.securesms.database.DatabaseFactory.upgradeRestored(DatabaseFactory.java:138)
06-15 19:29:39.529 23385 23414 E AndroidRuntime: 	at org.thoughtcrime.securesms.RegistrationActivity$2.doInBackground(RegistrationActivity.java:376)
06-15 19:29:39.529 23385 23414 E AndroidRuntime: 	at org.thoughtcrime.securesms.RegistrationActivity$2.doInBackground(RegistrationActivity.java:364)
06-15 19:29:39.546  4893  5829 W ActivityManager:   Force finishing activity org.thoughtcrime.securesms/.RegistrationActivity
06-15 19:29:39.567  4893  5829 W ActivityManager:   Force finishing activity org.thoughtcrime.securesms/.RegistrationActivity
06-15 19:29:39.568  4893  5829 W ActivityManager: Duplicate finish request for ActivityRecord{3c1c0de u0 org.thoughtcrime.securesms/.RegistrationActivity t3487 f}
06-15 19:30:09.133  4893  5829 I ActivityManager: Killing 23385:org.thoughtcrime.securesms/u0a113 (adj 900): crash

And here the log of trying to start Signal after the installation/crash:

06-15 19:30:40.394  4893  5829 I ActivityManager: START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10200000 cmp=org.thoughtcrime.securesms/.RoutingActivity (has extras)} from uid 10029 on display 0
06-15 19:30:40.415  4893 11098 I ActivityManager: Start proc 23490:org.thoughtcrime.securesms/u0a113 for activity org.thoughtcrime.securesms/.RoutingActivity
06-15 19:30:40.752 23490 23490 W PersistentAlarmManagerListener: class org.thoughtcrime.securesms.service.RotateSignedPreKeyListener scheduling for: 1529252953456
06-15 19:30:40.756 23490 23490 W PersistentAlarmManagerListener: class org.thoughtcrime.securesms.service.DirectoryRefreshListener scheduling for: 1529123353470
06-15 19:30:40.761 23490 23490 W PersistentAlarmManagerListener: class org.thoughtcrime.securesms.service.UpdateApkRefreshListener scheduling for: 1529101753473
06-15 19:30:40.777 23490 23516 E SQLiteOpenHelper: 	at org.thoughtcrime.securesms.database.helpers.SQLCipherOpenHelper.onUpgrade(SQLCipherOpenHelper.java:172)
06-15 19:30:40.777 23490 23516 E SQLiteOpenHelper: 	at org.thoughtcrime.securesms.database.helpers.SQLCipherOpenHelper.getReadableDatabase(SQLCipherOpenHelper.java:191)
06-15 19:30:40.777 23490 23516 E SQLiteOpenHelper: 	at org.thoughtcrime.securesms.database.SmsDatabase.getExpirationStartedMessages(SmsDatabase.java:668)
06-15 19:30:40.777 23490 23516 E SQLiteOpenHelper: 	at org.thoughtcrime.securesms.service.ExpiringMessageManager$LoadTask.run(ExpiringMessageManager.java:57)
06-15 19:30:40.786 23490 23516 E AndroidRuntime: Process: org.thoughtcrime.securesms, PID: 23490
06-15 19:30:40.786 23490 23516 E AndroidRuntime: net.sqlcipher.database.SQLiteException: Can't upgrade read-only database from version 6 to 7: /data/user/0/org.thoughtcrime.securesms/databases/signal.db
06-15 19:30:40.786 23490 23516 E AndroidRuntime: 	at org.thoughtcrime.securesms.database.helpers.SQLCipherOpenHelper.getReadableDatabase(SQLCipherOpenHelper.java:191)
06-15 19:30:40.786 23490 23516 E AndroidRuntime: 	at org.thoughtcrime.securesms.database.SmsDatabase.getExpirationStartedMessages(SmsDatabase.java:668)
06-15 19:30:40.786 23490 23516 E AndroidRuntime: 	at org.thoughtcrime.securesms.service.ExpiringMessageManager$LoadTask.run(ExpiringMessageManager.java:57)
06-15 19:30:40.788  4893  5083 W ActivityManager:   Force finishing activity org.thoughtcrime.securesms/.RoutingActivity
06-15 19:30:40.922  4893  5355 I ActivityManager: START u0 {cmp=org.thoughtcrime.securesms/.RegistrationActivity (has extras)} from uid 10113 on display 0
06-15 19:30:40.935  4893  5829 W ActivityManager: Duplicate finish request for ActivityRecord{1bf314b u0 org.thoughtcrime.securesms/.RoutingActivity t3488 f}
06-15 19:30:40.966  4893  4927 I WindowManager: Failed to capture screenshot of Token{a6f0728 ActivityRecord{1bf314b u0 org.thoughtcrime.securesms/.RoutingActivity t3488 f}} appWin=Window{d796279 u0 Starting org.thoughtcrime.securesms} drawState=4
06-15 19:30:41.086  4893  4930 W BroadcastQueue: Skipping deliver [background] BroadcastRecord{cbc967e u0 org.thoughtcrime.securesms.service.action.NEW_KEY_EVENT} to ReceiverList{9d7312b 23490 org.thoughtcrime.securesms/10113/u0 remote:a93857a}: process crashing
06-15 19:30:41.107 23490 23521 E SQLiteOpenHelper: 	at org.thoughtcrime.securesms.database.helpers.SQLCipherOpenHelper.onUpgrade(SQLCipherOpenHelper.java:172)
06-15 19:30:41.107 23490 23521 E SQLiteOpenHelper: 	at org.thoughtcrime.securesms.database.helpers.SQLCipherOpenHelper.getReadableDatabase(SQLCipherOpenHelper.java:191)
06-15 19:30:41.107 23490 23521 E SQLiteOpenHelper: 	at org.thoughtcrime.securesms.database.MmsSmsDatabase.queryTables(MmsSmsDatabase.java:341)
06-15 19:30:41.107 23490 23521 E SQLiteOpenHelper: 	at org.thoughtcrime.securesms.database.MmsSmsDatabase.getUnread(MmsSmsDatabase.java:130)
06-15 19:30:41.107 23490 23521 E SQLiteOpenHelper: 	at org.thoughtcrime.securesms.notifications.MessageNotifier.updateNotification(MessageNotifier.java:239)
06-15 19:30:41.107 23490 23521 E SQLiteOpenHelper: 	at org.thoughtcrime.securesms.notifications.MessageNotifier.updateNotification(MessageNotifier.java:190)
06-15 19:30:41.107 23490 23521 E SQLiteOpenHelper: 	at org.thoughtcrime.securesms.service.KeyCachingService$1.doInBackground(KeyCachingService.java:115)
06-15 19:30:41.107 23490 23521 E SQLiteOpenHelper: 	at org.thoughtcrime.securesms.service.KeyCachingService$1.doInBackground(KeyCachingService.java:111)
06-15 19:30:41.141  4893 11096 I ActivityManager: Process org.thoughtcrime.securesms (pid 23490) has died
06-15 19:30:41.143  4893 11096 W ActivityManager: Scheduling restart of crashed service org.thoughtcrime.securesms/.service.KeyCachingService in 1000ms
06-15 19:30:41.169  4893 11096 I ActivityManager: Start proc 23523:org.thoughtcrime.securesms/u0a113 for activity org.thoughtcrime.securesms/.RegistrationActivity
06-15 19:30:41.172  4893  5083 W ActivityManager: Ignoring remove of inactive process: ProcessRecord{8879c8a 0:org.thoughtcrime.securesms/u0a113}
06-15 19:30:41.347 23523 23523 W PersistentAlarmManagerListener: class org.thoughtcrime.securesms.service.RotateSignedPreKeyListener scheduling for: 1529252953456
06-15 19:30:41.351 23523 23523 W PersistentAlarmManagerListener: class org.thoughtcrime.securesms.service.DirectoryRefreshListener scheduling for: 1529123353470
06-15 19:30:41.353 23523 23523 W PersistentAlarmManagerListener: class org.thoughtcrime.securesms.service.UpdateApkRefreshListener scheduling for: 1529101753473
06-15 19:30:41.365 23523 23548 E SQLiteOpenHelper: 	at org.thoughtcrime.securesms.database.helpers.SQLCipherOpenHelper.onUpgrade(SQLCipherOpenHelper.java:172)
06-15 19:30:41.365 23523 23548 E SQLiteOpenHelper: 	at org.thoughtcrime.securesms.database.helpers.SQLCipherOpenHelper.getReadableDatabase(SQLCipherOpenHelper.java:191)
06-15 19:30:41.365 23523 23548 E SQLiteOpenHelper: 	at org.thoughtcrime.securesms.database.SmsDatabase.getExpirationStartedMessages(SmsDatabase.java:668)
06-15 19:30:41.365 23523 23548 E SQLiteOpenHelper: 	at org.thoughtcrime.securesms.service.ExpiringMessageManager$LoadTask.run(ExpiringMessageManager.java:57)
06-15 19:30:41.374 23523 23548 E AndroidRuntime: Process: org.thoughtcrime.securesms, PID: 23523
06-15 19:30:41.374 23523 23548 E AndroidRuntime: net.sqlcipher.database.SQLiteException: Can't upgrade read-only database from version 6 to 7: /data/user/0/org.thoughtcrime.securesms/databases/signal.db
06-15 19:30:41.374 23523 23548 E AndroidRuntime: 	at org.thoughtcrime.securesms.database.helpers.SQLCipherOpenHelper.getReadableDatabase(SQLCipherOpenHelper.java:191)
06-15 19:30:41.374 23523 23548 E AndroidRuntime: 	at org.thoughtcrime.securesms.database.SmsDatabase.getExpirationStartedMessages(SmsDatabase.java:668)
06-15 19:30:41.374 23523 23548 E AndroidRuntime: 	at org.thoughtcrime.securesms.service.ExpiringMessageManager$LoadTask.run(ExpiringMessageManager.java:57)
06-15 19:30:41.377  4893  4928 W ActivityManager: Process org.thoughtcrime.securesms has crashed too many times: killing!
06-15 19:30:41.378  4893  4928 W ActivityManager:   Force finishing activity org.thoughtcrime.securesms/.RegistrationActivity
06-15 19:30:41.432  4893  4928 E ActivityManager: Found activity ActivityRecord{7dac2d0 u0 org.thoughtcrime.securesms/.RegistrationActivity t3489 f} in proc activity list using null instead of expected ProcessRecord{e2319a9 23523:org.thoughtcrime.securesms/u0a113}
06-15 19:30:41.450  4893  4928 I ActivityManager: Killing 23523:org.thoughtcrime.securesms/u0a113 (adj 0): crash

**Comment by anon1701** [comment]
Short update regarding the problem: same crashing happens with different backup files, so the problem does not seem to be in single backup file.

**Comment by tbsn** [comment]
I'm not sure if this is related, but the occurrence of crashes, the affected versions and the error-line `net.sqlcipher.database.SQLiteException: Can't upgrade read-only database from version 6 to 7: /data/user/0/org.thoughtcrime.securesms/databases/signal.db` are mentioned here as well: https://github.com/signalapp/Signal-Android/issues/7694#issuecomment-385372506

At least you might have luck importing your backup on a fresh 4.18.3 installation.

**Comment by anon1701** [comment]
@tbsn thank you for the tip.
Fresh 4.18.3 installation was able to restore the backup and is running nicely. Something in 4.19.8 seems to causing the problem.

**Comment by greyson-signal** [comment]
@anon1701 4.20.8 is live on production now. Do you mind checking to see if that solves your problems? There were a couple backup-related fixes in that release.

**Comment by anon1701** [comment]
@greyson-signal I would be happy to test 4.20.8, but it is not available for download from Signals web page yet (https://signal.org/android/apk still offers version 4.19.3).
Let me know when I can download 4.20.8 and I will test it.

**Comment by RiseT** [comment]
@anon1701 4.21 eems to be up now.

**Comment by firestorm99** [comment]
Just tried to restore a backup created in Signal 4.21.6 on an other phone with Signal 4.21.6. After entering the code it starts restoring(counter goes up), but at some point it just closes. 

Backup size: 195.07MB
Backup created on: Xiaomi Mi 5C, Android 7.1.2, latest Stable 9.5.2
Tried to restore on: Huawei P20 Lite, Android 8.0.0, Build 8.0.0.136 (latest)

Logcat: 

`06-28 22:01:16.401  8071  8671 E AndroidRuntime: FATAL EXCEPTION: AsyncTask #4
06-28 22:01:16.401  8071  8671 E AndroidRuntime: Process: org.thoughtcrime.securesms, PID: 8071
06-28 22:01:16.401  8071  8671 E AndroidRuntime: java.lang.RuntimeException: An error occurred while executing doInBackground()
06-28 22:01:16.401  8071  8671 E AndroidRuntime: 	at android.os.AsyncTask$3.done(AsyncTask.java:365)
06-28 22:01:16.401  8071  8671 E AndroidRuntime: 	at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
06-28 22:01:16.401  8071  8671 E AndroidRuntime: 	at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
06-28 22:01:16.401  8071  8671 E AndroidRuntime: 	at java.util.concurrent.FutureTask.run(FutureTask.java:271)
06-28 22:01:16.401  8071  8671 E AndroidRuntime: 	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:257)
06-28 22:01:16.401  8071  8671 E AndroidRuntime: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
06-28 22:01:16.401  8071  8671 E AndroidRuntime: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
06-28 22:01:16.401  8071  8671 E AndroidRuntime: 	at java.lang.Thread.run(Thread.java:784)
06-28 22:01:16.401  8071  8671 E AndroidRuntime: Caused by: java.lang.NegativeArraySizeException: -1627120993
06-28 22:01:16.401  8071  8671 E AndroidRuntime: 	at org.thoughtcrime.securesms.backup.FullBackupImporter$BackupRecordInputStream.readFrame(FullBackupImporter.java:303)
06-28 22:01:16.401  8071  8671 E AndroidRuntime: 	at org.thoughtcrime.securesms.backup.FullBackupImporter$BackupRecordInputStream.readFrame(FullBackupImporter.java:245)
06-28 22:01:16.401  8071  8671 E AndroidRuntime: 	at org.thoughtcrime.securesms.backup.FullBackupImporter.importFile(FullBackupImporter.java:79)
06-28 22:01:16.401  8071  8671 E AndroidRuntime: 	at org.thoughtcrime.securesms.RegistrationActivity$2.doInBackground(RegistrationActivity.java:372)
06-28 22:01:16.401  8071  8671 E AndroidRuntime: 	at org.thoughtcrime.securesms.RegistrationActivity$2.doInBackground(RegistrationActivity.java:364)
06-28 22:01:16.401  8071  8671 E AndroidRuntime: 	at android.os.AsyncTask$2.call(AsyncTask.java:345)
06-28 22:01:16.401  8071  8671 E AndroidRuntime: 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
06-28 22:01:16.401  8071  8671 E AndroidRuntime: 	... 4 more
06-28 22:01:16.405  1155  1657 W ActivityManager:   Force finishing activity org.thoughtcrime.securesms/.RegistrationActivity`

**Comment by anon1701** [comment]
Installed Signal version 4.21.6 (downloaded from https://signal.org/android/apk) and the same issue remains.
`````
07-02 18:39:16.404 22650 22677 E SQLiteOpenHelper: net.sqlcipher.database.SQLiteException: duplicate column name: quote_id: ALTER TABLE mms ADD COLUMN quote_id INTEGER DEFAULT 0
07-02 18:39:16.404 22650 22677 E SQLiteOpenHelper: 	at net.sqlcipher.database.SQLiteDatabase.native_execSQL(Native Method)
07-02 18:39:16.404 22650 22677 E SQLiteOpenHelper: 	at net.sqlcipher.database.SQLiteDatabase.execSQL(SQLiteDatabase.java:2266)
07-02 18:39:16.404 22650 22677 E SQLiteOpenHelper: 	at net.sqlcipher.database.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:173)
07-02 18:39:16.404 22650 22677 E SQLiteOpenHelper: 	at net.sqlcipher.database.SQLiteOpenHelper.getReadableDatabase(SQLiteOpenHelper.java:227)
07-02 18:39:16.404 22650 22677 E SQLiteOpenHelper: 	at net.sqlcipher.database.SQLiteOpenHelper.getReadableDatabase(SQLiteOpenHelper.java:214)
07-02 18:39:16.428 22650 22677 E AndroidRuntime: net.sqlcipher.database.SQLiteException: Can't upgrade read-only database from version 6 to 10: /data/user/0/org.thoughtcrime.securesms/databases/signal.db
`````
The similiar error message/behaviour can be observer on here #7945.

I've never installed a self-compiled version of Signal. Is there any simple way to restore to 4.20.9 version of Signal? I would not want to start from a fresh install.
I'm happy to provide any additional logs etc.

**Comment by tbsn** [comment]
@anon1701 based on the log and the addition in your last comment I suspect it might be the same problem as in the issue I linked earlier and it's solved [here](https://community.signalusers.org/t/update-chrashes-how-to-edit-signal-db/3256).

**Comment by anon1701** [comment]
@tbsn I think you are correct, this most probably is the same problem. I do not have any experience with compiling Android software. Any ideas how I could get the database fixed?
Would it be possible for you to upload the "database fix"-version .apk somewhere where I could download it?

**Comment by tbsn** [comment]
@anon1701 I'm not comfortable with making software publicly available that potentially will mess things up, if not used for this specific purpose. Also, I generally would not recommend to install critical software from "someoneontheinternet"!
If this nevertheless would be the only solution to you, you can drop me your mail over on the forum.

**Comment by anon1701** [comment]
@tbsn you are correct. I followed your instructions and was able to restore the backup properly to the newest non-google version of Signal.
Thank you a lot, I would not have had a clue how to proceed without your advice.

@greyson-signal as an idea to prevent this kind of errors happening in the future, could there be some kind of check of database/Signal version that will prevent problems when going restoring a backup that has been made with a newer version of Signal? I could imagine that this will not happen too often, but at least I found personally find out one such scenario, since the non-google version of Signal often drags behind a bit in versioning compared to the google-enabled version.
